src/Uno.UI/UI/Xaml/UIElement.Layout.netstd.cs

```diff
@@ -4,98 +4,204 @@
+		private protected void ClearLayoutFlags(LayoutFlag flag) => _layoutFlags &= ~flag;
+
+		/// <summary>
+		/// Reset flags to original state
+		/// </summary>
+		[MethodImpl(MethodImplOptions.AggressiveInlining)]
+		private protected void ResetLayoutFlags() => _layoutFlags = DEFAULT_STARTING_LAYOUTFLAGS;
+
+		internal bool IsMeasureDirty
+		{
+			[MethodImpl(MethodImplOptions.AggressiveInlining)]
+			get => GetLayoutFlag(LayoutFlag.MeasureDirty);
+		}
+
+		internal bool IsMeasureDirtyPath
+		{
+			[MethodImpl(MethodImplOptions.AggressiveInlining)]
+			get => GetLayoutFlag(LayoutFlag.MeasureDirtyPath);
+		}
+
+		internal bool IsMeasureOrMeasureDirtyPath
+		{
+			[MethodImpl(MethodImplOptions.AggressiveInlining)]
+			get => GetAnyFlag(LayoutFlag.MeasureDirty | LayoutFlag.MeasureDirtyPath);
+		}
+
+		internal bool IsArrangeDirty
+		{
+			[MethodImpl(MethodImplOptions.AggressiveInlining)]
+			get => GetLayoutFlag(LayoutFlag.ArrangeDirty);
+		}
+
 		public void InvalidateMeasure()
 		{
 			if (ShouldInterceptInvalidate)
 			{
 				return;
 			}
 
-			// TODO: Figure out why this condition breaks layouting in some cases
-			//if (_isMeasureValid)
+			if (IsMeasureDirty)
 			{
-				_isMeasureValid = false;
-				_isArrangeValid = false;
-				if (this.GetParent() is UIElement parent)
-				{
-					parent.InvalidateMeasure();
-				}
-				else
+				return; // already dirty
+			}
+
+			SetLayoutFlag(LayoutFlag.MeasureDirty);
+
+			if (FeatureConfiguration.UIElement.UseInvalidateMeasurePath)
+			{
+				InvalidateParentMeasurePath();
+			}
+			else
+			{
+				(this.GetParent() as UIElement)?.InvalidateMeasure();
+				if (IsVisualTreeRoot)
 				{
 					Window.InvalidateMeasure();
 				}
 			}
 		}
 
+		[MethodImpl(MethodImplOptions.AggressiveInlining)]
+		private void InvalidateMeasurePath()
+		{
+			if(IsMeasureOrMeasureDirtyPath)
+			{
+				return; // Already invalidated
+			}
+
+			SetLayoutFlag(LayoutFlag.MeasureDirtyPath);
+
+			InvalidateParentMeasurePath();
+		}
+
+		[MethodImpl(MethodImplOptions.AggressiveInlining)]
+		internal void InvalidateParentMeasurePath()
+		{
+			if (this.GetParent() is UIElement parent)
+			{
+				parent.InvalidateMeasurePath();
+			}
+			else if (IsVisualTreeRoot)
+			{
+				Window.InvalidateMeasure();
+			}
+		}
+
 		public void InvalidateArrange()
 		{
 			if (ShouldInterceptInvalidate)
 			{
 				return;
 			}
 
-			if (_isArrangeValid)
+			if (IsArrangeDirty)
 			{
-				_isArrangeValid = false;
-				if (this.GetParent() is UIElement parent)
-				{
-					parent.InvalidateArrange();
-				}
-				else
-				{
-					Window.InvalidateMeasure();
-				}
+				return; // Already dirty
+			}
+
+			SetLayoutFlag(LayoutFlag.ArrangeDirty);
+			if (this.GetParent() is UIElement parent)
+			{
+				parent.InvalidateArrange();
+			}
+			else
+			{
+				Window.InvalidateArrange();
 			}
 		}
 
 		public void Measure(Size availableSize)
 		{
-			if (!(this is FrameworkElement))
+			if (this is not FrameworkElement)
```

Comment:
We already have a flag `_isFrameworkElement` in netstd which is computed only once in ctor

Reply:
One more _actual_ flag opportunity ;)

Reply:
I saw it and I decided to restrained myself from converting everything.